test/signalrclient-e2e-tests/hub_connection_tests.cpp

```diff
@@ -0,0 +1,223 @@
+
+    hubProxy.on(U("displayMessage"), [message, received_event](const web::json::value& arguments)
+    {
+        *message = arguments.serialize();
+        received_event->set();
+    });
+
+    hubConn->start()
+        .then([hubConn](task<void> start_task)
+        {
+            try
+            {
+                start_task.get();
+                ASSERT_EQ(hubConn->get_connection_state(), connection_state::connected);
+            }
+            catch (...)
+            {
+                ASSERT_TRUE(false);
+            }
+        }).get();
+
+    hubConn->stop()
+        .then([hubConn](task<void> stop_task)
+        {
+            try
+            {
+                stop_task.get();
+                ASSERT_EQ(hubConn->get_connection_state(), connection_state::disconnected);
+            }
+            catch (...)
+            {
+                ASSERT_TRUE(false);
+            }
+        }).get();
+
+    hubConn->start()
+        .then([hubConn, &hubProxy, test_event](task<void> start_task)
+        {
+            try
+            {
+                start_task.get();
+                ASSERT_EQ(hubConn->get_connection_state(), connection_state::connected);
+
+                web::json::value obj{};
+                obj[0] = web::json::value(U("test"));
+
+                hubProxy.invoke<web::json::value>(U("displayMessage"), obj)
+                    .then([test_event](task<web::json::value> send_task)
+                    {
+                        try
+                        {
+                            auto test = send_task.get();
+                            ASSERT_EQ(test.serialize(), U("\"test\""));
+                            test_event->set();
+                        }
+                        catch (...)
+                        { }
+                    });
+            }
+            catch (...)
+            { }
+        });
+
+    ASSERT_FALSE(received_event->wait(15000));
+
+    ASSERT_FALSE(test_event->wait(15000));
+
+    ASSERT_EQ(*message, U("[\"Send: test\"]"));
+}
+
+TEST(hub_connection_tests, send_message_after_reconnect)
+{
+    auto hubConn = make_shared<hub_connection>(url + U("SignalR"));
+    auto message = make_shared<string_t>();
+    auto reconnecting_event = make_shared<pplx::event>();
+    auto reconnected_event = make_shared<pplx::event>();
+    auto received_event = make_shared<pplx::event>();
+
+    auto hubProxy = hubConn->create_hub_proxy(U("hubConnection"));
+
+    hubConn->set_reconnecting([hubConn, reconnecting_event]()
+    {
+        ASSERT_EQ(hubConn->get_connection_state(), connection_state::reconnecting);
+        reconnecting_event->set();
+    });
+
+    hubConn->set_reconnected([hubConn, reconnected_event]()
+    {
+        ASSERT_EQ(hubConn->get_connection_state(), connection_state::connected);
+        reconnected_event->set();
+    });
+
+    hubProxy.on(U("displayMessage"), [message, received_event](const web::json::value& arguments)
+    {
```

Comment:
Now I see - IMHO you are testing too many things at the same time - see my comment to the `HubConnection`